frontends/web/src/hooks/keyboard.ts

```diff
@@ -71,59 +71,132 @@ export const useFocusTrap = (
   ref: React.RefObject<HTMLElement>,
   active: boolean,
 ) => {
+  const autofocusDelay = 50;
+
   const previouslyFocused = useRef<HTMLElement | null>(null);
+  const trapEnabled = useRef(false);
+  const autofocusTimer = useRef<number | null>(null);
+  const cancelledAutofocus = useRef(false);
 
-  // Focus trap handler
-  const handleKeyDown = (e: KeyboardEvent) => {
-    if (!active || !ref.current || e.key !== 'Tab') {
-      return;
-    }
-    const node = ref.current;
-    const focusables = node.querySelectorAll<HTMLElement>(FOCUSABLE_SELECTOR);
-    if (!focusables.length) {
-      return;
-    }
+  const handleKeyDown = useCallback(
+    (e: KeyboardEvent) => {
+      if (!trapEnabled.current || !ref.current || e.key !== 'Tab') {
+        return;
+      }
 
-    const first = focusables[0];
-    const last = focusables[focusables.length - 1];
+      const focusables = Array.from(
+        ref.current.querySelectorAll<HTMLElement>(FOCUSABLE_SELECTOR)
+      ).filter((el) => el.offsetParent !== null); // skip hidden
 
-    if (e.shiftKey && document.activeElement === first) {
-      e.preventDefault();
-      last?.focus();
-    } else if (!e.shiftKey && document.activeElement === last) {
-      e.preventDefault();
-      first?.focus();
-    }
-  };
+      if (focusables.length === 0) {
+        return;
+      }
 
-  useKeydown(handleKeyDown);
+      const first = focusables[0];
+      const last = focusables[focusables.length - 1];
+      const current = document.activeElement as HTMLElement;
+
+      if (e.shiftKey && current && current === first) {
```

Comment:
I added `&& current` as here both `first` and `current` could be `undefined`. And we don't want to continue in case of `undefined === undefined`.